src/providers/MeetingProvider/MeetingManager.ts

```diff
@@ -463,6 +444,22 @@ export class MeetingManager implements AudioVideoObserver {
     this.publishDeviceLabelTriggerChange();
   };
 
+  setConstraints = (deviceLabels: DeviceLabels) => {
```

Comment:
We cannot tell what kind of devices builder want to ask for permission in their customized `deviceLabelTrigger`.

Expose this function so when user passes the `deviceLabels: DeviceLabelTrigger` to `meetingManager.join()`, they can set the value of `constraints` to align it.

Reply:
This is a breaking API change — a working app will stop working unless they now call this method, right?

This is all an indication that this design is incorrect.

When the hook runs, it gets a stream. The stream's tracks will include audio and/or video. Can we use that knowledge?

Reply:
> This is a breaking API change — a working app will stop working unless they now call this method, right?

I don't think this is a breaking API, there are 4 cases for `deviceLabelTrigger` (same for the customized `deviceLabelTrigger`):

| Case | DeviceLabelTrigger                                                      | No `setConstraints` calling                | With  `setConstraints` calling |
| ---- | ----------------------------------------------------------------------- | ------------------------------------------ | ------------------------------ |
| 1.   | `async () => await navigator.getUserMedia({audio: true});`              | First ask for mic, then camera (incorrect) | Only one for mic(fixed)        |
| 2.   | `async () => await navigator.getUserMedia({video: true});`              | First ask for camera, then mic (incorrect) | Only one for camera(fixed)     |
| 3.   | `async () => await navigator.getUserMedia({audio: true, video: true});` | Only one for mic and camera                | Only one for mic and camera    |
| 4.   | `() => Promise.resolve(new MediaStream());`                             | None                                       | None                           |

With calling `setConstraints` a working App will still work as before, while a buggy App now get fixed.

> When the hook runs, it gets a stream. The stream's tracks will include audio and/or video. Can we use that knowledge?

I don't think we can use that. In case 1 and 2:
1. The first device permission prompt is triggered by `audioVideo.list*Devices()` -> `audioVideo.listDevicesOfKind()` -> `deviceLabelTrigger`.
    * The stream we get. We can't use this to determine the below one, or it could be a big change.
2. The second device permission prompt is triggered by `audioVide.choose*Device(deviceId)`;

Reply:
What's the behavior for a customer's app upgrading from an older version of the component library?

If I understand correctly, what you're observing here is that the browser is remembering the requested permission, and so the issue is that the request for the real media stream (chooseDevice) triggers a second permission prompt if it is not a subset of a previously requested permission. That's actually OK.

The behavior in the table you wrote above should depend on whether you're calling `chooseAudioDevice` or `chooseVideoDevice`.

I think this is going in the wrong direction: if a customer specifies a hook, it means they want control.

If I use a device label trigger of `() => navigator.getUserMedia({audio: true})` and then call `chooseAudioInputDevice`, what is causing a camera request to occur?